airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,279 @@
+          primary_key_sorter = primary_key_sorter + ", "
+        primary_key_sorter = primary_key_sorter + f"{key} ASC"
+      return primary_key_sorter
+
+    def generate_ordered_query(self, stream_slice):
+      values = ', '.join(self.properties) # we use values instead of '*' because we want to preserve the order of the columns
+      incremental_column_sorter = f", {self.incremental_column} ASC" if self.incremental_column else ""
+      # per https://docs.oracle.com/en/cloud/saas/netsuite/ns-online-help/section_156257805177.html#subsect_156330163819
+      # date literals need to be wrapped in a to_date function
+      incremental_column_filter = f"{self.incremental_column} >= to_timestamp('{stream_slice['first_day']}', 'YYYY-MM-DD') AND {self.incremental_column} <= to_timestamp('{stream_slice['last_day']}', 'YYYY-MM-DD')" if self.incremental_column else ""
+      primary_key_filter = self.generate_primary_key_filter()
+      primary_key_sorter = self.generate_primary_key_sorter()
+      and_connector = " AND " if primary_key_filter != "" and incremental_column_filter != "" else ""
+      where_clause = "WHERE " if primary_key_filter != "" or incremental_column_filter != "" else ""
+      query = f"""
+        SELECT TOP {NETSUITE_PAGINATION_INTERVAL} {values} FROM {self.table_name} 
+        {where_clause}{primary_key_filter}{and_connector}{incremental_column_filter}
+        ORDER BY {primary_key_sorter}""" + incremental_column_sorter
+      return query
+  
+    def serialize_row_to_response(self, row):
+      response = {}
+      for i, column in enumerate(self.properties):
+        response[column] = row[i]
+      return response
+
+    def stream_slices(
+      self, *, sync_mode: SyncMode, cursor_field: Optional[List[str]] = None, stream_state: Optional[Mapping[str, Any]] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+      # return [{'first_day': date(2000, 1, 1), 'last_day': date(2023, 12,31)}]
+      start_date, end_date = self.get_range_to_fetch(sync_mode=sync_mode, stream_state=stream_state)
+
+      # Ensure start_date is before end_date
+      if start_date > end_date:
+        start_date, end_date = end_date, start_date
+
+      current_date = start_date
+      date_slices = []
+      while current_date <= end_date:
+        date_slices.append(self.get_slice_for_year(current_date))
+        current_date = current_date.replace(year=current_date.year + 1, month=1)
+
+      return date_slices
+    
+    def get_range_to_fetch(self, sync_mode: SyncMode, stream_state: Optional[Mapping[str, Any]] = None):
+      end_slice_range = date.today()
+      # we fetch data for some years in the future because accountants can book transactions in the future
+      end_slice_range = end_slice_range.replace(year=end_slice_range.year + YEARS_FORWARD_LOOKING, day=1) 
+      # we use an earliest date of 1980 since 1998 was when netsuite was founded.  Note that in some cases customers might have ported over older historical data
```

Comment:
should the start date be an optional config field with a default value of 1980?

Reply:
I could also see this being useful if the user is only interested in recent data, but I don't know if that's a real use case

Reply:
Customers definitely had different historical data requirements at Rutter.  I think we should have this be an optional config field.

Reply:
Added this option.  resolving